atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/CassandraKeyValueServiceConfigsTest.java

```diff
@@ -69,24 +84,120 @@ public void canDeserialize() throws IOException, URISyntaxException {
+        assertThatThrownBy(() -> CONFIGS_WITH_KEYSPACE.copyWithResolvedKeyspace(Optional.of(KEYSPACE_2)))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void copyWithResolvedKeyspaceResolvesConfigWithOriginalKeyspaceIfNoNamespaceProvided() {
+        CassandraKeyValueServiceConfigs newConfigs = CONFIGS_WITH_KEYSPACE.copyWithResolvedKeyspace(Optional.empty());
+        assertThat(newConfigs.installConfig().getKeyspaceOrThrow())
+                .isEqualTo(CONFIG_WITH_KEYSPACE.getKeyspaceOrThrow());
+    }
+
+    @Test
+    public void copyWithResolvedKeyspaceResolvesConfigWithNamespaceIfNoKeyspaceProvided() {
+        CassandraKeyValueServiceConfigs newConfigs =
+                CONFIGS_WITHOUT_KEYSPACE.copyWithResolvedKeyspace(Optional.of(KEYSPACE));
+        assertThat(newConfigs.installConfig().getKeyspaceOrThrow()).isEqualTo(KEYSPACE);
     }
 
     @Test
-    public void otherPropertiesConservedWhenAddingKeyspace() {
-        CassandraKeyValueServiceConfig newConfig =
-                CassandraKeyValueServiceConfigs.copyWithKeyspace(CONFIG_WITHOUT_KEYSPACE, KEYSPACE);
-        assertThat(newConfig.credentials()).isEqualTo(CREDENTIALS);
+    public void copyWithResolvedKeyspaceResolvesConfigIfKeyspaceAndNamespaceProvidedAndMatch() {
+        CassandraKeyValueServiceConfigs newConfigs =
+                CONFIGS_WITH_KEYSPACE.copyWithResolvedKeyspace(Optional.of(KEYSPACE));
+        assertThat(newConfigs.installConfig().getKeyspaceOrThrow()).isEqualTo(KEYSPACE);
     }
 
     @Test
-    public void canReplaceKeyspace() {
-        CassandraKeyValueServiceConfig newConfig =
-                CassandraKeyValueServiceConfigs.copyWithKeyspace(CONFIG_WITH_KEYSPACE, KEYSPACE_2);
-        assertThat(newConfig.getKeyspaceOrThrow()).isEqualTo(KEYSPACE_2);
+    public void fromKeyValueServiceConfigMergesInstallAndRuntime() {
+        CassandraKeyValueServiceConfigs configs = CassandraKeyValueServiceConfigs.fromKeyValueServiceConfigsOrThrow(
+                CONFIG_WITHOUT_KEYSPACE,
+                Refreshable.only(Optional.of(ImmutableCassandraKeyValueServiceRuntimeConfig.builder()
+                        .replicationFactor(1010101)
+                        .build())));
+        assertThat(configs.runtimeConfig().get().replicationFactor())
+                .isEqualTo(configs.installConfig().replicationFactor().orElseThrow());
+    }
+
+    @Test
+    public void emptyRuntimeConfigShouldResolveToDefaultRuntimeConfig() {
+        CassandraKeyValueServiceConfigs returnedConfigs =
+                CassandraKeyValueServiceConfigs.fromKeyValueServiceConfigsOrThrow(
+                        CONFIG_WITHOUT_KEYSPACE, Refreshable.only(Optional.empty()));
+
+        assertThat(returnedConfigs.runtimeConfig().get().mutationBatchCount())
+                .isEqualTo(CassandraKeyValueServiceRuntimeConfig.getDefault().mutationBatchCount());
```

Comment:
Can we check more fields of this than just mutation batch count? This seems possibly errorprone as-is.

Reply:
Sure, I can actually just test that the configs are equal.